core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -176,189 +193,188 @@ oppia.factory('NumberWithUnitsObjectFactory', [
-    var unitList = [];
-    for (var key in unitDict) {
-      unitList.push({unit: key, exponent: unitDict[key]});
-    }
-    return unitList;
-  };
-
-  var unitToList = function(unitsWithMultiplier) {
-    var unitDict = {};
-    for (var i = 0; i < unitsWithMultiplier.length; i++) {
-      var unit = unitsWithMultiplier[i][0];
-      var multiplier = unitsWithMultiplier[i][1];
-      var ind = unit.indexOf('^');
-      if (ind > -1) {
-        var s = unit.substr(0, ind);
-        var power = parseInt(unit.substr(ind + 1));
-      } else {
-        var s = unit;
-        var power = 1;
-      }
-      if (!(s in unitDict)) {
-        unitDict[s] = 0;
+      return convertUnitDictToList(unitDict);
+    };
+
+    Units.prototype.toDict = function() {
+      return {
+        units: this.units
+      };
+    };
+
+    Units.fromList = function(unitsList) {
+      return new Units(unitsList);
+    };
+
+    Units.fromStringToList = function(unitsString) {
+      return unitToList(unitWithMultiplier(Units.stringToLexical(unitsString)));
+    };
+
+    Units.prototype.toString = function() {
+      var unit = '';
+      for (var i = 0; i < this.units.length; i++) {
+        var d = this.units[i];
+        if (d.exponent === 1) {
+          unit += d.unit + ' ';
+        } else {
+          unit += d.unit + '^' + d.exponent.toString() + ' ';
+        }
       }
-      unitDict[s] += multiplier * power;
-    }
-    return convertUnitDictToList(unitDict);
-  };
-
-  Units.prototype.toDict = function() {
-    return {
-      units: this.units
+      return unit.trim();
     };
-  };
-
-  Units.fromList = function(unitsList) {
-    return new Units(unitsList);
-  };
-
-  Units.fromStringToList = function(unitsString) {
-    return unitToList(unitWithMultiplier(Units.stringToLexical(unitsString)));
-  };
-
-  Units.prototype.toString = function() {
-    var unit = '';
-    for (var i = 0; i < this.units.length; i++) {
-      var d = this.units[i];
-      if (d.exponent === 1) {
-        unit += d.unit + ' ';
-      } else {
-        unit += d.unit + '^' + d.exponent.toString() + ' ';
+
+    Units.createCurrencyUnits = function() {
+      // Creates user-defined currency (base + sub) units.
+      math.createUnit('dollar', {aliases: CURRENCY_UNITS.dollar});
+      math.createUnit('cent', {
+        definition: '0.01 dollar', aliases: CURRENCY_UNITS.cent});
+      math.createUnit('rupee', {aliases: CURRENCY_UNITS.rupee});
+      math.createUnit('paise', {
+        definition: '0.01 rupee', aliases: CURRENCY_UNITS.paise});
+    };
+
+    Units.toMathjsCompatibleString = function(units) {
+      // Makes the units compatible with the math.js allowed format.
+      units = units.replace(/per/g, '/');
+
+      // Special symbols need to be replaced as math.js doesn't support custom
+      // units starting with special symbols. Also, it doesn't allow units
+      // followed by a number as in the case of currency units.
+      if (units.includes('$')) {
```

Comment:
This whole part should also be handled through currency list.

Reply:
done